ReactiveCocoa/Swift/Signal.swift

```diff
@@ -135,19 +146,28 @@ public final class Signal<Value, Error: ErrorProtocol> {
 	///   - observer: An observer to forward the events to.
 	///
 	/// - returns: An optional `Disposable` which can be used to disconnect the
-	///            observer. Disposing of the Disposable will have no effect on
-	///            the Signal itself.
+	///            observer.
 	@discardableResult
 	public func observe(_ observer: Observer) -> Disposable? {
 		var token: RemovalToken?
-		atomicObservers.modify { observers in
-			token = observers?.insert(observer)
+		state.modify { state in
+			if state != nil {
+				if state!.observers.isEmpty {
```

Comment:
This `isEmpty` check isn't really necessary since setting `state.retainedSignal` again has no effect.